File: clients/android/app/src/main/java/eu/allodia/mailcal/MailboxEmpty.kt
package eu.allodia.mailcal

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import uniffi.mailcal_bindings.EmptyReason

/**
 * Why the mail list has no rows, drawn over the rows themselves.
 *
 * The badge beside the folder counts what the server holds over all time, while the list can only
 * show what sync depth kept (`docs/folder-pane.md`, rule 5). A folder of older mail then badges its
 * unread above nothing at all, and without this nothing on screen reconciles the two numbers.
 *
 * Which case it is comes from the core, so this composable picks no wording of its own; the action
 * appears only where widening the depth would actually find something.
 */
@Composable
internal fun MailboxEmpty(
    reason: EmptyReason?,
    onOpenSettings: () -> Unit,
    modifier: Modifier = Modifier,
) {
    if (reason == null) return
    val ctx = LocalContext.current
    Column(
        modifier = modifier.fillMaxSize().padding(horizontal = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        when (reason) {
            is EmptyReason.NoMail -> Text(
                text = L10n.mailbox_empty_no_mail(ctx),
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )
            is EmptyReason.OutsideSyncDepth -> {
                Text(
                    text = L10n.mailbox_empty_windowed(ctx, reason.months.toInt()),
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center,
                )
                Text(
                    text = L10n.mailbox_empty_windowed_body(ctx),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(top = 8.dp),
                )
                TextButton(onClick = onOpenSettings, modifier = Modifier.padding(top = 8.dp)) {
                    Text(L10n.mailbox_empty_change(ctx))
                }
            }
        }
    }
}

File: clients/linux/src/ui/mailbox_empty.rs
//! Why the mail list has no rows, drawn in the list's own placeholder.
//!
//! The badge beside the folder counts what the server holds over all time, while the list can
//! only show what sync depth kept (`docs/folder-pane.md`, rule 5). A folder of older mail then
//! badges its unread above nothing at all, and without this nothing on screen reconciles the two
//! numbers.
//!
//! A `GtkListBox` placeholder is shown exactly when the box holds no rows, so the rendering path
//! beside this one never has to ask whether it is drawing an empty list: it renders rows, and GTK
//! decides which of the two the user sees.

use adw::prelude::*;
use mailcal_bindings::EmptyReason;

use super::AppInput;
use crate::l10n;

/// The icon for a folder whose mail is simply not here yet, and for one the depth is keeping out.
const EMPTY_ICON: &str = "mail-mark-unread-symbolic";
const WINDOWED_ICON: &str = "document-open-recent-symbolic";

/// Puts the reason for an empty list into `list`'s placeholder, or takes the placeholder away
/// when the core gave no reason (a list with rows, or a search, which says how far it looked
/// through the horizon strip instead).
pub(super) fn render(
    list: &gtk::ListBox,
    reason: Option<&EmptyReason>,
    sender: &relm4::Sender<AppInput>,
) {
    let Some(reason) = reason else {
        list.set_placeholder(None::<&gtk::Widget>);
        return;
    };
    let page = adw::StatusPage::new();
    match reason {
        EmptyReason::NoMail => {
            // Nothing was held back, so there is no setting to offer: a button here would
            // promise mail that widening cannot find.
            page.set_icon_name(Some(EMPTY_ICON));
            page.set_title(l10n::mailbox_empty_no_mail());
        }
        EmptyReason::OutsideSyncDepth { months } => {
            page.set_icon_name(Some(WINDOWED_ICON));
            page.set_title(&l10n::mailbox_empty_windowed(i64::from(*months)));
            page.set_description(Some(l10n::mailbox_empty_windowed_body()));
            // A statement the user cannot act on is half the value, so the page carries the way
            // to the depth setting, as the search horizon strip does.
            let change = gtk::Button::with_label(l10n::mailbox_empty_change());
            change.add_css_class("pill");
            change.add_css_class("suggested-action");
            change.set_halign(gtk::Align::Center);
            let input = sender.clone();
            change.connect_clicked(move |_| input.emit(AppInput::OpenSyncDepthSettings));
            page.set_child(Some(&change));
        }
    }
    list.set_placeholder(Some(&page));
}
